[SPARK-7451][YARN] Preemption of executors is counted as failure causing Spark job to fail

ashwinshankar77 opened the pull request:
Added a check to handle container exit status for the preemption scenario, log an INFO message in such cases and move on.
@andrewor14

Commit:
Stop counting preemption of executors as failure

AmplabJenkins commented:
Can one of the admins verify this patch?

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala:
```diff
@@ -373,7 +373,11 @@ private[yarn] class YarnAllocator(
         // there are some exit status' we shouldn't necessarily count against us, but for
         // now I think its ok as none of the containers are expected to exit
-        if (completedContainer.getExitStatus == -103) { // vmem limit exceeded
+        if (completedContainer.getExitStatus == ContainerExitStatus.PREEMPTED) {
```
Does this constant exist all the way back to Hadoop 2.2? IIRC preemption was added later.

tgravescs replied:
the exit status has it back in 2.1.1-beta

ashwinshankar77 replied:
@vanzin yes its been there since 2.1.1-beta. Here is the JIRA which added that constant: https://issues.apache.org/jira/browse/YARN-1049

sryza commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala:
```diff
@@ -373,7 +373,11 @@ private[yarn] class YarnAllocator(
         // now I think its ok as none of the containers are expected to exit
-        if (completedContainer.getExitStatus == -103) { // vmem limit exceeded
+        if (completedContainer.getExitStatus == ContainerExitStatus.PREEMPTED) {
+          logInfo("Container killed: " + containerId +
```
Can this say "Container preempted?"  Also, we shouldn't need to report the diagnostics and exit status because they will always be the same for preemption.

tgravescs commented:
Note that the AllocateResponse also contains PreemptionMessage which is list of containers RM is asking for back.  AM could potentially use this to decide which containers to kill.  That can be future enhancement though.

I think this looks good other then @sryza  comments.

Commit:
Fix log info message

ashwinshankar77 commented:
Thanks all for reviewing !
@sryza Addressed your comment in the latest patch.
@tgravescs Yes, one use case on top of my mind is try to avoid preempting containers
which has RDD's cached. However, I believe the current implementation in the scheduler
doesn't take inputs from AM for preemption and that the list of containers in PreemptionMessage
is more of a notification to AM, no ?

tgravescs commented:
Yeah I'm sure it would require more plumbing hence a future enhancement if we decide its useful.

sryza commented:
LGTM

andrewor14 commented:
Looks like we never ran tests for this. Jenkins test this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32272 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32272/consoleFull) for   PR 5993 at commit [`90900cf`](https://github.com/apache/spark/commit/90900cf23da0a82394f534d4c28fa30f1e860630).

andrewor14 commented:
LGTM2

SparkQA commented:
[Test build #32272 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32272/consoleFull) for   PR 5993 at commit [`90900cf`](https://github.com/apache/spark/commit/90900cf23da0a82394f534d4c28fa30f1e860630).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32272/
Test PASSed.

sryza commented:
Committing this

ashwinshankar77 commented:
Thanks for committing this, Sandy !
